File: Progetto Map Estensione A.A. 2024-2025/app/src/main/java/com/example/qtereshold/network/SocketQtClient.java
package com.example.qtereshold.network;

import java.io.Closeable;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.ConnectException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;

public class SocketQtClient implements Closeable {

    private final String host;
    private final int port;
    private final int timeoutMs;

    private Socket socket;
    private ObjectOutputStream out;
    private ObjectInputStream in;

    public SocketQtClient(String host, int port) {
        this(host, port, 10_000); // Default timeout
    }

    public SocketQtClient(String host, int port, int timeoutMs) {
        this.host = host;
        this.port = port;
        this.timeoutMs = timeoutMs;
    }

    /**
     * Si connette al server.
     * @throws IOException se la connessione fallisce.
     */
    public void connect() throws IOException {
        if (socket != null && socket.isConnected() && !socket.isClosed()) {
            return;
        }
        try {
            Socket s = new Socket();
            s.connect(new InetSocketAddress(host, port), timeoutMs);
            s.setSoTimeout(timeoutMs); // Timeout anche sulla lettura
            socket = s;
            out = new ObjectOutputStream(s.getOutputStream());
            out.flush();
            in = new ObjectInputStream(s.getInputStream());
        } catch (ConnectException e) {
            throw new IOException("Server offline o porta " + port + " chiusa", e);
        } catch (SocketTimeoutException e) {
            throw new IOException("Timeout di connessione: nessuna risposta dal server", e);
        } catch (UnknownHostException e) {
            throw new IOException("Host non valido: " + host, e);
        }
    }

    /**
     * Invia un comando e un parametro, riceve una singola stringa.
     * @param command Il comando (es. 10, 12, 13)
     * @param param Il parametro (es. nome tabella, nome file)
     * @return La stringa di risposta dal server
     * @throws IOException se la scrittura/lettura fallisce
     * @throws ClassNotFoundException se la risposta non è una Stringa
     */
    private String simpleRequestResponse(int command, Object param) throws IOException, ClassNotFoundException {
        if (out == null || in == null) throw new IOException("Client non connesso.");

        out.writeObject(command);
        out.writeObject(param);
        out.flush();
        return (String) in.readObject();
    }

    // --- Metodi del Protocollo Semplice ---

    public String storeTableFromDb_Simple(String table) throws IOException, ClassNotFoundException {
        return simpleRequestResponse(10, table);
    }

    public String learningFromDbTable_Simple(double radius) throws IOException, ClassNotFoundException {
        return simpleRequestResponse(11, radius);
    }

    public String saveClustersToFile_Simple(String filename) throws IOException, ClassNotFoundException {
        return simpleRequestResponse(12, filename);
    }

    public String learningFromFile_Simple(String filename) throws IOException, ClassNotFoundException {
        return simpleRequestResponse(13, filename);
    }

    public void ping() throws IOException, ClassNotFoundException {
        if (out == null || in == null) throw new IOException("Client non connesso.");

        out.writeObject(99);
        out.flush();
        String response = (String) in.readObject();
        if (!"PONG".equals(response)) {
            throw new IOException("Risposta Heartbeat non valida dal server.");
        }
    }

    // --- Metodi di utilità ---

    public boolean isConnected() {
        return socket != null && socket.isConnected() && !socket.isClosed();
    }

    public void disconnect() {
        close();
    }

    @Override
    public void close() {
        try { if (in != null) in.close(); } catch (Exception e) {}
        try { if (out != null) out.close(); } catch (Exception e) {}
        try { if (socket != null) socket.close(); } catch (Exception e) {}
    }
}

File: Progetto Map Estensione A.A. 2024-2025/app/src/main/java/com/example/qtereshold/network/SocketQtClient.kt
package com.example.qtereshold.network

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.Closeable
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.net.InetSocketAddress
import java.net.Socket
import java.io.IOException
import java.net.*

class SocketQtClient(
    private val host: String,
    private val port: Int = 8080,
    private val timeoutMs: Int = 10_000
) : Closeable {

    private var socket: Socket? = null
    private var out: ObjectOutputStream? = null
    private var `in`: ObjectInputStream? = null

    suspend fun connect() = withContext(Dispatchers.IO) {
        if (socket?.isConnected == true && socket?.isClosed == false) return@withContext
        try {
            val s = Socket()
            s.connect(InetSocketAddress(host, port), timeoutMs)
            s.soTimeout = timeoutMs
            socket = s
            out = ObjectOutputStream(s.getOutputStream())
            out!!.flush()
            `in` = ObjectInputStream(s.getInputStream())
        } catch (e: ConnectException) {
            throw IOException("Server offline o porta $port chiusa", e)
        } catch (e: SocketTimeoutException) {
            throw IOException("Timeout di connessione: nessuna risposta dal server", e)
        } catch (e: UnknownHostException) {
            throw IOException("Host non valido: $host", e)
        }
    }

    // ===================================================================
    // --- PROTOCOLLO SEMPLICE (per Client Android) ---
    // ===================================================================

    /**
     * Invia il COMANDO 10 (protocollo semplice).
     * Si aspetta UNA SOLA stringa in risposta ("OK" o "Errore: ...")
     */
    suspend fun storeTableFromDb_Simple(table: String): String = withContext(Dispatchers.IO) {
        try {
            requireNotNull(out).writeObject(10)
            requireNotNull(out).flush()
            requireNotNull(out).writeObject(table)
            requireNotNull(out).flush()
            // Legge la singola risposta
            requireNotNull(`in`).readObject() as String
        } catch (e: Exception) {
            "Errore: ${e.message ?: e.javaClass.simpleName}"
        }
    }

    /**
     * Invia il COMANDO 11 (protocollo semplice).
     * Si aspetta UNA SOLA stringa in risposta (i cluster o "Errore: ...")
     */
    suspend fun learningFromDbTable_Simple(radius: Double): String = withContext(Dispatchers.IO) {
        try {
            requireNotNull(out).writeObject(11)
            requireNotNull(out).flush()
            requireNotNull(out).writeObject(radius)
            requireNotNull(out).flush()
            // Legge la singola risposta
            requireNotNull(`in`).readObject() as String
        } catch (e: Exception) {
            "Errore: ${e.message ?: e.javaClass.simpleName}"
        }
    }

    /**
     * Invia il COMANDO 12 (protocollo semplice).
     * Si aspetta UNA SOLA stringa in risposta ("OK" o "Errore: ...")
     */
    suspend fun saveClustersToFile_Simple(filename: String): String = withContext(Dispatchers.IO) {
        try {
            requireNotNull(out).writeObject(12)
            requireNotNull(out).flush()
            requireNotNull(out).writeObject(filename)
            requireNotNull(out).flush()
            // Legge la singola risposta
            requireNotNull(`in`).readObject() as String
        } catch (e: Exception) {
            "Errore: ${e.message ?: e.javaClass.simpleName}"
        }
    }

    /**
     * Invia il COMANDO 13 (protocollo semplice).
     * Si aspetta UNA SOLA stringa in risposta (i cluster o "Errore: ...")
     */
    suspend fun learningFromFile_Simple(filename: String): String = withContext(Dispatchers.IO) {
        try {
            requireNotNull(out).writeObject(13)
            requireNotNull(out).flush()
            requireNotNull(out).writeObject(filename)
            requireNotNull(out).flush()
            // Legge la singola risposta
            requireNotNull(`in`).readObject() as String
        } catch (e: Exception) {
            "Errore: ${e.message ?: e.javaClass.simpleName}"
        }
    }

    /**
     * Invia il COMANDO 99 (Heartbeat/Ping) per controllare se la connessione è attiva.
     * Si aspetta "PONG" in risposta.
     * Lancia una IOException se la connessione è caduta.
     */
    suspend fun ping() = withContext(Dispatchers.IO) {
        // Blocco try-catch per intercettare qualsiasi errore (incluso un timeout
        // se il server è bloccato, o una SocketException se è chiuso)
        try {
            requireNotNull(out).writeObject(99)
            requireNotNull(out).flush()
            val response = requireNotNull(`in`).readObject() as String
            if (response != "PONG") {
                // Il server ha risposto in modo strano
                throw IOException("Risposta Heartbeat non valida dal server.")
            }
        } catch (e: Exception) {
            // Qualsiasi errore qui (SocketException, EOFException, ClassCastException)
            // significa che la connessione è morta. Rilanciamo l'eccezione
            // per farla gestire al chiamante (MainActivity).
            throw IOException("Heartbeat fallito: connessione persa.", e)
        }
    }


    // =================================================================
    // --- Funzioni di utilità ---
    // =================================================================

    fun isConnected(): Boolean = socket?.isConnected == true && socket?.isClosed == false

    fun disconnect() = close()

    override fun close() {
        try { `in`?.close() } catch (_: Exception) {}
        try { out?.close() } catch (_: Exception) {}
        try { socket?.close() } catch (_: Exception) {}
    }
}